projects/OctoberSky1/config/UOStateMachine.cpp

```diff
@@ -0,0 +1,271 @@
+	// InternalEvent(ST_FLIGHT);
+}
+
+// Exit action when ExitDescent state exits.
+EXIT_DEFINE(UOStateMachine, ExitDescent)
+{
+	std::cout << "RocketSM::ExitDescent\n";
+}
+
+// Entry action when ExitDescent state exits.
+ENTRY_DEFINE(UOStateMachine, EnterGround, UOSMData)
+{
+	enterNewState(ST_GROUND);
+
+	std::cout << "RocketSM::EnterGround\n";
+}
+
+// code for the ground state
+STATE_DEFINE(UOStateMachine, Ground, UOSMData)
+{
+	rocketInterface.update(data);
+	rocketData = rocketInterface.getLatest();
+
+	detectExternEvent(rocketData);
+	// showInfo(rocketData);
+}
+
+void UOStateMachine::detectExternEvent(const sensorsData *data)
+{
+	int eventNbr = data->inputEventNumber;
+	switch (eventNbr)
+	{
+	case 0:
+		Start();
+		break;
+	case 1:
+		Apogee();
+		break;
+	case 2:
+		Touchdown();
+		break;
+	default:
+		break;
+	}
+
+#if USE_SOCKET_CLIENT
+	eventNbr = data->clientEventNumber;
+	switch (eventNbr)
+	{
+	case 0:
+		Start();
+		break;
+	case 1:
+		Apogee();
+		break;
+	case 2:
+		Touchdown();
+		break;
+	default:
+		break;
+	}
+#endif
+}
+
+void UOStateMachine::detectApogee(const sensorsData *data)
+{
+#if USE_SBG
+	// TODO: only check for apogee x seconds after launch
+	// Euler angle
+	// pitch is Yangle
+	static uint8_t consecutiveEvents = 0;
+
+	float Yangle = (180 / PI) * (acos(cos(data->sbg.Xangle * (PI / 180)) * cos(data->sbg.Yangle * (PI / 180))));
+
+	if (Yangle >= 45)
+	{
+		consecutiveEvents++;
+	}
+	else
+	{
+		consecutiveEvents = 0;
+	}
+
+	// trigger appogee if the sbg detects "ApogeeThreshold" number of consecutive times
+	// that the rocket is pointing downwards and falling
+	if (consecutiveEvents >= ApogeeThreshold)
+	{
+		std::cout << "Apogee \n";
+		Apogee();
+	}
+#endif
+}
+
+void UOStateMachine::showInfo(const sensorsData *data)
+{
+#if USE_SBG
+	printf("Barometer: %f\tGps: longitude %f\t latitude %f\t altitude %f\t Velocity: N %f\tE %f\tD %f\tSolutionStatus %d\t%d\n",
+		   data->sbg.barometricAltitude,
+		   data->sbg.gpsLatitude, data->sbg.gpsLongitude, data->sbg.gpsAltitude,
+		   data->sbg.velocityN, data->sbg.velocityE, data->sbg.velocityD,
+		   data->sbg.solutionStatus,
+		   (data->sbg.solutionStatus) & 0b1111);
+// std::cout << std::bitset<32>(data->sbg.solutionStatus) << "\n";
+#endif
+	return;
+}
+
+void UOStateMachine::enterNewState(States state)
+{
+	StateMachine::enterNewState(state);
+}
+
+double UOStateMachine::getValueForTime(double minimum, double maximum, duration_ms targetTime)
+{
+	duration_ns timeSinceEntry = std::chrono::steady_clock::now() - entryTime;
+	double progress = ((double)timeSinceEntry.count()) / duration_ns(targetTime).count();
+	return std::min(maximum, minimum + progress * (maximum - minimum));
+}
+
+bool UOStateMachine::switchStatesAfterTime(States state, duration_ms targetTime)
+{
+	duration_ns timeSinceEntry = std::chrono::steady_clock::now() - entryTime;
+	if (timeSinceEntry >= duration_ns(targetTime))
+	{
+		InternalEvent(state);
+
+		return true;
+	}
+
+	return false;
+}
```

Comment:
Is there a reason you aren't using the function from StateMachine.cpp?

Reply:
Me and Seb didn't know that the functions were transferred in State Machine.cpp so we copied these implementations from past commits. I'll remove them and use the ones in StateMachine.cpp.